x/eibc/keeper/lps.go

```diff
@@ -0,0 +1,236 @@
+		return errorsmod.Wrap(err, "set by rollapp denom")
+	}
+	return nil
+}
+
+func (s LPs) Get(ctx sdk.Context, id uint64) (*types.OnDemandLPRecord, error) {
+	ret, err := s.byID.Get(ctx, id)
+	return &ret, err
+}
+
+// reason is human-readable string for debugging/ux
+func (s LPs) Del(ctx sdk.Context, id uint64, reason string) error {
+	lp, err := s.byID.Get(ctx, id)
+	if err != nil {
+		return errorsmod.Wrap(err, "get id")
+	}
+	err = s.byID.Remove(ctx, id)
+	if err != nil {
+		return errorsmod.Wrap(err, "remove id")
+	}
+	err = s.byRollAppDenom.Remove(ctx, collections.Join3(lp.Lp.Rollapp, lp.Lp.Denom, id))
+	if err != nil {
+		return errorsmod.Wrap(err, "remove by rollapp denom")
+	}
+	err = s.byAddr.Remove(ctx, collections.Join(lp.Lp.FundsAddr, lp.Id))
+	if err != nil {
+		return errorsmod.Wrap(err, "remove by addr")
+	}
+	if err := uevent.EmitTypedEvent(ctx, &types.EventDeletedOnDemandLP{
+		Id:        id,
+		FundsAddr: lp.Lp.FundsAddr,
+		Reason:    reason,
+	}); err != nil {
+		return errorsmod.Wrap(err, "event")
+	}
+	return nil
+}
+
+func (s LPs) GetByAddr(ctx sdk.Context, addr sdk.AccAddress) ([]*types.OnDemandLPRecord, error) {
+	var ret []*types.OnDemandLPRecord
+	rng := collections.NewPrefixedPairRange[string, uint64](addr.String())
+	iter, err := s.byAddr.Iterate(ctx, rng)
+	if err != nil {
+		return nil, err
+	}
+	for ; iter.Valid(); iter.Next() {
+		key, err := iter.Key()
+		if err != nil {
+			return nil, err
+		}
+		id := key.K2()
+		lp, err := s.byID.Get(ctx, id)
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, &lp)
+	}
+	return ret, err
+}
+
+func (s LPs) GetOrderCompatibleLPs(ctx sdk.Context, o types.DemandOrder) ([]types.OnDemandLPRecord, error) {
+	rol := o.RollappId
+	denom := o.Denom()
+	ranger := collections.NewSuperPrefixedTripleRange[string, string, uint64](rol, denom)
+	iter, err := s.byRollAppDenom.Iterate(ctx, ranger)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Close()
+
+	var compat []types.OnDemandLPRecord
+	for ; iter.Valid(); iter.Next() {
+		key, err := iter.Key()
+		if err != nil {
+			return nil, err
+		}
+		id := key.K3()
+
+		lpr, err := s.byID.Get(ctx, id)
+		if err != nil {
+			return nil, err
+		}
+		if lpr.Accepts(uint64(ctx.BlockHeight()), o) {
+			compat = append(compat, lpr)
+		}
+	}
+	return compat, nil
+}
+
+func (k Keeper) FulfillByOnDemandLP(ctx sdk.Context, order string, rng int64) error {
+	o, err := k.GetOutstandingOrder(ctx, order)
+	if err != nil {
+		return errorsmod.Wrap(err, "get outstanding order")
+	}
+	lps, err := k.LPs.GetOrderCompatibleLPs(ctx, *o)
+	if err != nil {
+		return errorsmod.Wrap(err, "get compatible lp")
+	}
+	r := rand.New(rand.NewSource(rng))
+	r.Shuffle(len(lps), func(i, j int) {
+		lps[i], lps[j] = lps[j], lps[i]
+	})
+	for _, lp := range lps {
+		err := k.Fulfill(ctx, o, lp.Lp.MustAddr())
+		if err != nil {
+			if errorsmod.IsOf(err, sdkerrors.ErrInsufficientFunds) {
+				if err := k.LPs.Del(ctx, lp.Id, "out of funds"); err != nil {
```

Comment:
wouldn't it be more elegant to first filter down the lps by their funds, to make sure they have enough for the order?

Reply:
I think the current approach is a bit lazier, in the happy case it's less work.
Your way is good too, but I don't think the difference in elegance is significant.

I'll keep as is for now iydm